--- /opt/SDL2/android-project/src/org/libsdl/app/SDLActivity.java	2013-07-31 14:08:39.000000000 +0900
+++ SDLActivity.java	2013-09-20 17:44:31.000000000 +0900
@@ -53,10 +53,11 @@
     // Load the .so
     static {
         System.loadLibrary("SDL2");
-        //System.loadLibrary("SDL2_image");
+        System.loadLibrary("SDL2_image");
         //System.loadLibrary("SDL2_mixer");
         //System.loadLibrary("SDL2_net");
         //System.loadLibrary("SDL2_ttf");
+        System.loadLibrary("pxsrc");
         System.loadLibrary("main");
     }
 
@@ -434,11 +435,11 @@
         try {
             EGL10 egl = (EGL10)EGLContext.getEGL();
 
-            egl.eglWaitNative(EGL10.EGL_CORE_NATIVE_ENGINE, null);
+            // egl.eglWaitNative(EGL10.EGL_CORE_NATIVE_ENGINE, null);
 
             // drawing here
 
-            egl.eglWaitGL();
+            // egl.eglWaitGL();
 
             egl.eglSwapBuffers(SDLActivity.mEGLDisplay, SDLActivity.mEGLSurface);
 
